File: src/simulation/culture/thinking/language/templates/TextInfo.java
package simulation.culture.thinking.language.templates;

import simulation.Controller;
import simulation.culture.thinking.meaning.Meme;

import java.util.HashMap;
import java.util.Map;

public class TextInfo {
    private Map<String, Meme> map = new HashMap<>();

    public TextInfo(Meme actor, Meme verb, Meme receiver) {
        map.put("!actor", actor);
        map.put("@verb", verb);
        map.put("!receiver", receiver);
    }

    public Map<String, Meme> getMap() {
        return map;
    }

    public Meme getMainPart(String key) {
        return getSubstituted(key).topMemeCopy();
    }

    public Meme getSubstituted(String key) {
        return substitute(map.get(key));
    }

    public Meme substitute(Meme meme) {
        return meme.refactor(m -> {
            if (Controller.session.templateBase.templateChars.contains(m.getObserverWord().charAt(0))) {
                Meme substitution = map.get(m.getObserverWord());
                if (substitution == null) {
                    throw new RuntimeException();
                }
                return substitution.copy();
            } else {
                return m.topMemeCopy();
            }
        });
    }
}

File: src/simulation/culture/thinking/language/templates/TextInfo.kt
package simulation.culture.thinking.language.templates

import simulation.Controller
import simulation.culture.thinking.meaning.Meme

class TextInfo(val map: MutableMap<InfoKey, Meme>) {
    constructor(actor: Meme, verb: Meme, receiver: Meme) : this(mutableMapOf()) {
        map["!actor"] = actor
        map["@verb"] = verb
        map["!receiver"] = receiver
    }

    fun changedInfo(key: String, newMeme: Meme): TextInfo {
        val changedMap = map.toMutableMap()
        changedMap[key] = newMeme
        return TextInfo(changedMap)
    }

    fun getMainPart(key: InfoKey) = getSubstituted(key).topMemeCopy()

    fun getSubstituted(key: InfoKey) = substitute(map.getValue(key))

    fun substitute(meme: Meme): Meme {
        return meme.refactor {
            if (Controller.session.templateBase.templateChars.contains(it.observerWord[0])) {
                val substitution = map[it.observerWord] ?: throw RuntimeException()
                return@refactor substitution.copy()
            } else {
                return@refactor it.topMemeCopy()
            }
        }
    }
}

typealias InfoKey = String